Approved/enhanced-array-literals.md

```diff
@@ -0,0 +1,251 @@
+If Q# requires that every type has a default value, then uninhabited types are not possible to express properly.
+This can even cause subtle bugs in the soundness of a type system, as demonstrated by [this bug in Java's generics](https://hackernoon.com/java-is-unsound-28c84cb2b3f) that happened because of the existence of a value (specifically `null`) for a type that should have been uninhabited.
+
+Removing default-initialized array constructors, and the requirement that every type has a default value, resolves these problems.
+
+## Enhancing array literals
+
+With the removal of default-initialized array constructors, alternatives are needed to create empty arrays and arrays with a repeated initial value.
+The alternative for `new T[n]` is `[x, size = n]`, where `x` is the initial value of each item.
+In the case where `n` = 0, the alternative for `new T[0]` is `[]`.
+
+# Description
+
+## Current Status
+
+Q# currently supports this syntax:
+
+```qsharp
+// Create an empty array of type Int[].
+let empty = new Int[0];
+
+// Create an array of length 10 of type Bool[] where every value is false.
+let bools = new Bool[10];
+```
+
+## Proposed Modification
+
+The syntax above will be deprecated and removed.
+It is replaced by the constructs described below.
+
+### Empty arrays
+
+The empty array literal, `[]`, will be supported.
+Unlike `new T[0]`, this literal does not specify the type of the array to be created.
+Its type will be inferred based on its first usage.
+
+```qsharp
+function SumI(xs : Int[]) : Int {
+    return Fold(PlusI, 0, xs);
+}
+
+function SumD(xs : Double[]) : Int {
+    return Fold(PlusD, 0.0, xs);
+}
+
+function Example() : Unit {
+    // The type of [] is inferred to be Int[] based on its position as the argument to SumI.
+    let x = SumI([]);
+
+    // The type of empty1 is inferred to be Int[] based on its position as the argument to SumI.
+    let empty1 = [];
+    let y = SumI(empty1);
+
+    // The following line contains a compilation error, because empty1 is of type Int[], not
+    // Double[].
+    let z = SumD(empty1);
+
+    // Since empty2 is not used, its type cannot be inferred.
+    // This is a compilation error.
+    let empty2 = [];
+
+    // While empty3 is used, its usage does not provide any information about its concrete type.
+    // This is a compilation error.
+    let empty3 = [];
+    Message($"Empty array: {empty3}");
+
+    // In cases where the type of [] cannot be inferred, either [x, size = 0] or
+    // Microsoft.Quantum.Arrays.EmptyArray can be used instead.
+    // The following code is valid.
+    let empty4 = [0, size = 0];
+    let empty5 = EmptyArray<Int>();
+    Message($"Empty Int arrays: {empty4}, {empty5}");
+}
+```
+
+### Repeated constant arrays
+
+The semantics of `[x, size = n]` are identical to `new T[n]`, except that the initial value is now given by `x`:
+
+1. `n` is an expression of type `Int`.
```

Comment:
This sentence will not be accurate in the future, since we expect to be able to provide an `Int` tuple of any arity to the `size` parameter. This type cannot be expressed in Q#'s type system, which concerns me.

Reply:
I think one way to fix this is to forbid things like `[0, size = (3, 2)]` to construct nested arrays. If we restrict this to multidimensional arrays, then the type of `size` is determined by the number of `#`s, so `#[0, size = (n, n)]`, `##[0, size = (n, n, n)]`, etc.

Reply:
To construct a nested array, I recommend the more verbose but more explicit syntax `[[0, size = _], size = _]`.

Reply:
@bettinaheim, can you clarify whether you want to support these examples from your comment here: https://github.com/microsoft/qsharp-language/pull/48#issuecomment-732601443

* `[0, size = (3, 2)]`
* `#[0, size = (4, 3, 2)]`

Or if we can enforce stricter typing like in my examples above?

* `[[0, size = _], size = _]`
*  `##[0, size = (4, 3, 2)]`

Reply:
Updated the proposal to recommend against overloading on tuple arity for `size`, and prefer syntax with unambiguous types.

Reply:
Thanks for bringing this up. That it should always be possible to translate the introduced construct to a function call is a very good point and I agree with ensuring that we limit its usage in the multidimensional case to ensure that.